test/framework/toolchain.py

```diff
@@ -381,6 +381,42 @@ def test_optarch_aarch64_heuristic(self):
         self.assertEqual(tc.options.options_map['optarch'], 'mcpu=cortex-a72.cortex-a53')
         self.assertTrue('-mcpu=cortex-a72.cortex-a53' in os.environ['CFLAGS'])
         self.modtool.purge()
+  
+    def test_compiler_dependent_optarch(self):
+        """Test whether specifying optarch on a per compiler basis works."""
+        flag_vars = ['CFLAGS', 'CXXFLAGS', 'FCFLAGS', 'FFLAGS', 'F90FLAGS']
+        for intelflags,intel_expanded_flags in [('intelflag','intelflag'),('GENERIC','xSSE2'),('','')]:
+            for gccflags,gcc_expanded_flags in  [('gccflag','gccflag'),('GENERIC','march=x86-64 -mtune=generic'),('','')]:
+                optarch_var = 'Intel:%s,GCC:%s' % (intelflags,gccflags)
+                build_options = {'optarch': optarch_var}
+                init_config(build_options=build_options)
+                for toolchain, toolchain_version in [('iccifort','2011.13.367'),('GCC','4.7.2'),('PGI','16.7-GCC-5.4.0-2.26')]:
+                    for enable in [True, False]:
```

Comment:
No, I mean creating 4 separate lists, and then letting Python create a list of 4-tuples with all possible options (seems like `itertools.product` does what you need).

Anyway, this is less important, since this is only a test where code style is less of an issue